### ISSUE TITLE
blog.cloudflare.com assertion fails after connection close

### ISSUE THREAD

**Issue by agrover** [mozilla/neqo#278]
cmdline: `RUST_BACKTRACE=1 RUST_LOG=neqo_transport::dump ./target/debug/neqo-client https://blog.cloudflare.com/`
result:

```
</html>
<FIN[0]>
0s576ms DEBUG [Client 0x7fff66571b50] pn=27 type=Short
  TX -> ConnectionClose { error_code: Application(0), frame_type: 0, reason_phrase: [107, 116, 104, 120, 98, 121, 101, 33] }
thread 'main' panicked at 'assertion failed: self.state_active()', neqo-http3/src/connection.rs:523:9
stack backtrace:
   0: std::panicking::default_hook::{{closure}}
   1: std::panicking::default_hook
   2: std::panicking::rust_panic_with_hook
   3: std::panicking::begin_panic
             at /builddir/build/BUILD/rustc-1.38.0-src/src/libstd/panicking.rs:411
   4: neqo_http3::connection::Http3Connection<E,T,H>::handle_connection_closing
             at neqo-http3/src/connection.rs:523
   5: neqo_http3::connection::Http3Connection<E,T,H>::check_connection_events
             at neqo-http3/src/connection.rs:366
   6: neqo_http3::connection::Http3Connection<E,T,H>::process_http3
             at neqo-http3/src/connection.rs:281
   7: neqo_http3::connection_client::Http3Client::process_http3
             at neqo-http3/src/connection_client.rs:243
   8: neqo_client::process_loop
             at neqo-client/src/main.rs:133
   9: neqo_client::client
             at neqo-client/src/main.rs:281
  10: neqo_client::main
             at neqo-client/src/main.rs:318
  11: std::rt::lang_start::{{closure}}
             at /builddir/build/BUILD/rustc-1.38.0-src/src/libstd/rt.rs:64
```

**Comment by hawkinsw** [comment]
Patch incoming.

**Comment by hawkinsw** [comment]
This happens because the `Http3Connection` sets its state to `Closing` in the `Http3Connection::close` function before it calls `Connection::close`. `Connection::close()` sets /its/ state to `Closing` which adds a `ConnectionEvent::StateChange` which, when handled, calls `Http3Connection::handle_connection_closing` which asserts whether its state is connected!

Got that? :-)

I think that there are two ways to handle this:
1. I was about to submit a fix that changes the definition of `Http3Connection::state_active` to include `Closing` as an active state. I actually think that's reasonable. Until the connection is closed, even if it's going to be closed, it is still active.
2. However, I went with an alternate: Not setting the `Http3Connection` state to `Closing` until the `Http3Connection::handle_connection_closing` function is called. If this is not the right way to handle it, please let me know and I'm happy to take a different path!

Thanks for all the great work on this implementation!

**Comment by ddragana** [comment]
Can you please do not change state_active. This state was meant to mean that you can receive data on streams. And also I want to set set Http3Connection state to Closing as soon as we know we are closing.
Please add Closing to the assertion:
`assert!(self.state_active() | self.state_closing())`

Thanks!!!

**Comment by hawkinsw** [comment]
Absolutely! That was the third option that I didn't explore. I didn't do it
because it seemed like cheating :-) Standby for a new version of the patch!

**Comment by hawkinsw** [comment]
Thanks again for your feedback, @ddragana! Updated patch is available.

**Comment by hawkinsw** [comment]
@agrover I think that we can close this now. Seems to work now!

**Comment by ddragana** [comment]
Fixed by #281
